qiskit_nature/problems/second_quantization/lattice/models/heisenberg_model.py

```diff
@@ -0,0 +1,103 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"The Heisenberg model"
+import logging
+from typing import Optional
+from fractions import Fraction
+from qiskit_nature.operators.second_quantization import SpinOp
+from qiskit_nature.problems.second_quantization.lattice.lattices import Lattice
+from .lattice_model import LatticeModel
+
+logger = logging.getLogger(__name__)
+
+
+class HeisenbergModel(LatticeModel):
+    """The Heisenberg model."""
+
+    def __init__(
+        self,
+        lattice: Lattice,
+        model_constants: Optional[dict] = None,
+        ext_magnetic_field: Optional[dict] = None,
```

Comment:
These should not be dictionaries.
If I am understanding things correctly, we have the following parameters:
- `J_x`, `J_y`, `J_z`
- `B_x`, `B_y`, `B_z`
- `h`

How about, `J` simply gets provided as a tuple of floats of length 3.
The same can be done for `B`. That said, `B` even appears to have bool values, to simply turn on/off `h` along the given axis. Doesn't this make `h` redundant if we can simply have `B` be another tuple of floats of length 3?

Furthermore, in the `IsingModel`, we deliberately kept `lattice` as the sole input argument to `__init__`. I think we could do the same here and then supply `@classmethod` "constructors" similar to the `IsingModel.from_uniform_parameters` which would inject `J` and `B`, respectively. In that way, we could even have a method which simply provides `h` that gets expanded into the proposed tuple `B`.

Reply:
I agree with your suggestions. I wrote the inputs as dictionaries, because I did not know how I would inject them using the lattice. But now, with your explanation, I'm starting to see a way to do this.

I will update the code according with #746 and I will try to rewrite the class in this new way.